pkg/doc/jwt/jwt.go

```diff
@@ -0,0 +1,271 @@
+	return v(joseHeaders, payload, signingInput, signature)
+}
+
+func verifyUnsecuredJWT(joseHeaders jose.Headers, _, _, signature []byte) error {
+	alg, ok := joseHeaders.Algorithm()
+	if !ok {
+		return errors.New("alg is not defined")
+	}
+
+	if alg != AlgorithmNone {
+		return errors.New("alg value is not 'none'")
+	}
+
+	if len(signature) > 0 {
+		return errors.New("not empty signature")
+	}
+
+	return nil
+}
+
+// UnsecuredJWTVerifier provides verifier for unsecured JWT.
+func UnsecuredJWTVerifier() jose.SignatureVerifier {
+	return signatureVerifierFunc(verifyUnsecuredJWT)
+}
+
+type unsecuredJWTSigner struct {
+	extraHeaders map[string]interface{}
+}
+
+func (s unsecuredJWTSigner) Sign(_ []byte) ([]byte, error) {
+	return []byte(""), nil
+}
+
+func (s unsecuredJWTSigner) Headers() jose.Headers {
+	jHeaders := map[string]interface{}{
+		jose.HeaderAlgorithm: AlgorithmNone,
+		jose.HeaderType:      TypeJWT,
+	}
+
+	for k, v := range s.extraHeaders {
+		if _, ok := jHeaders[k]; !ok {
+			jHeaders[k] = v
+		}
+	}
+
+	return jHeaders
+}
+
+// JSONWebToken defines JSON Web Token (https://tools.ietf.org/html/rfc7519)
+type JSONWebToken struct {
+	Headers jose.Headers
+
+	Payload map[string]interface{}
+
+	signature []byte
+}
+
+// Parse parses input JWT in serialized form into JSON Web Token.
+// Currently JWS and unsecured JWT is supported.9
+func Parse(jwt string, opts ...ParseOpt) (*JSONWebToken, error) {
+	if !jose.IsCompactJWS(jwt) {
+		return nil, errors.New("JWT of compacted JWS form is supported only")
+	}
+
+	pOpts := &parseOpts{}
+
+	for _, opt := range opts {
+		opt(pOpts)
+	}
+
+	return parseJWS(jwt, pOpts)
+}
+
+// DecodeClaims fills input c with claims of a token.
+func (j *JSONWebToken) DecodeClaims(c interface{}) error {
+	pBytes, err := json.Marshal(j.Payload)
+	if err != nil {
+		return err
+	}
+
+	return json.Unmarshal(pBytes, c)
+}
+
+// LookupStringHeader makes look up of particular header with string value.
+func (j *JSONWebToken) LookupStringHeader(name string) string {
+	if headerValue, ok := j.Headers[name]; ok {
+		if headerStrValue, ok := headerValue.(string); ok {
+			return headerStrValue
+		}
+	}
+
+	return ""
+}
+
+// SerializeSigned makes (compact) serialization of token.
+func (j *JSONWebToken) SerializeSigned(signer jose.Signer, detached bool) (string, error) {
+	return j.serialize(signer, detached)
+}
+
+// SerializeUnsecured build unsecured JWT.
+func (j *JSONWebToken) SerializeUnsecured(extraHeaders map[string]interface{}, detached bool) (string, error) {
+	return j.serialize(&unsecuredJWTSigner{extraHeaders}, detached)
+}
+
+func (j *JSONWebToken) serialize(signer jose.Signer, detached bool) (string, error) {
+	payloadBytes, err := j.marshalPayload()
+	if err != nil {
+		return "", fmt.Errorf("marshal JWT claims: %w", err)
+	}
+
+	jws := jose.NewJWS(j.Headers, nil, payloadBytes)
+	j.signature = jws.Signature()
```

Comment:
So this seems a bit odd currently. When I reviewed this code previously, I thought signature wasn't populated by the constructor ... and hence this would end up returning nil?

Reply:
not relevant anymore as `JSONWebToken` is made immutable.